test/morgan.js

```diff
@@ -1316,6 +1317,102 @@ describe('morgan.compile(format)', function () {
   })
 })
 
+describe('morgan.token(name, function)', function () {
+  it('should overwrite existing functions if called multiple times', function (done) {
+    function token (req, res, arg) {
+      return arg && url.parse(req.url)[arg]
+    }
+    var urlToken = morgan.url
+    morgan.token('url', token)
+    assert.equal(morgan.url, token)
+    morgan.token('url', urlToken)
+    done()
+  })
+
+  describe('should use the string `-` if return value of the token function is falsey', function () {
+    function test (v, done) {
+      morgan.token('ret', function ret (req, res, arg) {
+        return arg
+      })
+      var cb = after(2, function (err, res, line) {
+        if (err) return done(err)
+        assert.equal(line, '-')
+        done()
+      })
+
+      var stream = createLineStream(function (line) {
+        cb(null, null, line)
+      })
+
+      var server = createServer(':ret', { stream: stream }, function (req, res, next) {
+        next()
+      })
+
+      request(server)
+      .get('/')
+      .expect(200, cb)
+    }
+    [
+      0,
+      NaN,
+      false,
+      '',
+      null,
+      undefined
+    ].forEach(function (v) {
+      it('for ' + (String(v) || JSON.stringify(v)), function (done) {
+        test(v, done)
+      })
+    })
+  })
+
+  it('should not see empty brackets as an argument value', function (done) {
+    morgan.token('checkempty', function ret (req, res, arg) {
+      assert.equal(arg, undefined)
+      assert.equal(arguments.length, 2)
+      cb(null)
+    })
+    var cb = after(3, function (err, res, line) {
+      if (err) return done(err)
+      assert.equal(line, '-[]')
+      done()
+    })
+
+    var stream = createLineStream(function (line) {
+      cb(null, null, line)
+    })
+
+    var server = createServer(':checkempty[]', { stream: stream }, function (req, res, next) {
+      next()
+    })
+
+    request(server)
+    .get('/')
+    .expect(200, cb)
+  })
+
+  it('should use the result', function (done) {
```

Comment:
This test says "morgan.token(name, function) should use the result". I'm not certain what that means. What should use the result? The result of what?